File: archive/e/elixir/baklava.ex
defmodule Baklava do
  @doc """
  Create baklava where the top has 10 spaces and then 1 asterisk.
  """
  @spec baklava() :: String.t()
  def baklava, do: baklava_grow(10, 1)

  @doc """
  Recursively grow the baklava until spaces <= zero.
  """
  @spec baklava_grow(spaces :: integer, asterisks :: integer) :: String.t()
  def baklava_grow(spaces, asterisks) when spaces <= 0 do
    line(0, asterisks) <> "\n" <> baklava_shrink(1, asterisks - 2)
  end

  def baklava_grow(spaces, asterisks) do
    line(spaces, asterisks) <> "\n" <> baklava_grow(spaces - 1, asterisks + 2)
  end

  @doc """
  Recursively shrink the baklava until asterisks <= zero.
  """
  @spec baklava_shrink(spaces :: integer, asterisks :: integer) :: String.t()
  def baklava_shrink(spaces, asterisks) when asterisks <= 1, do: line(spaces, 1)

  def baklava_shrink(spaces, asterisks) do
    line(spaces, asterisks) <> "\n" <> baklava_shrink(spaces + 1, asterisks - 2)
  end

  @doc """
  Return a single line of the baklava.
  """
  @spec line(spaces :: integer, asterisks :: integer) :: String.t()
  def line(spaces, asterisks) do
    String.duplicate(" ", spaces) <> String.duplicate("*", asterisks)
  end
end

Baklava.baklava() |> IO.puts()

File: archive/e/erlang/baklava.erl
-module(baklava).
-export([start/1]).

-spec start(MaxWidth :: integer()) -> ok.
start(MaxWidth) ->
    baklava(MaxWidth, 0),
    ok.

-spec baklava(MaxWidth :: integer(),
              Incrementor :: integer()) -> ok.
%%--------------------------------------------------------------------
%% Recursively prints the top half of the baklava 
%%--------------------------------------------------------------------
baklava(MaxWidth, MaxWidth) ->
    print((MaxWidth*2)+1, star),
    reverse_baklava(MaxWidth, MaxWidth-1);
baklava(MaxWidth, Incrementor) ->
    print(MaxWidth-Incrementor, space),
    print((Incrementor*2)+1, star),
    baklava(MaxWidth, Incrementor+1).

-spec reverse_baklava(MaxWidth :: integer(),
                      Decrementor :: integer()) -> ok.
%%--------------------------------------------------------------------
%% Recursively prints the bottom half of the baklava
%%--------------------------------------------------------------------
reverse_baklava(MaxWidth, 0) ->
    print(MaxWidth, space),
    print(1, star);
reverse_baklava(MaxWidth, Decrementor) ->
    print(MaxWidth-Decrementor, space),
    print((Decrementor*2)+1, star),
    reverse_baklava(MaxWidth, Decrementor-1).

-spec print(Number :: integer(),
            Type :: star | space) -> ok.
%%--------------------------------------------------------------------
%% Recursively prints the specified Number of stars or spaces
%%--------------------------------------------------------------------
print(0,star) ->
    io:format("~n");
print(0,_) ->
    ok;
print(N, space) ->
    io:format(" "),
    print(N-1, space);
print(N, star) ->
    io:format("*"),
    print(N-1, star).
